pkg/cloud/openstack/services/networking/loadbalancer.go

```diff
@@ -0,0 +1,438 @@
+	"sigs.k8s.io/cluster-api/pkg/util"
+	"time"
+)
+
+func (s *Service) ReconcileLoadBalancer(clusterName string, clusterProviderSpec *providerv1.OpenstackClusterProviderSpec, clusterProviderStatus *providerv1.OpenstackClusterProviderStatus) error {
+
+	if clusterProviderSpec.ExternalNetworkID == "" {
+		klog.V(3).Info("No need to create loadbalancer, due to missing ExternalNetworkID")
+		return nil
+	}
+	if clusterProviderSpec.APIServerLoadBalancerFloatingIP == "" {
+		klog.V(3).Info("No need to create loadbalancer, due to missing APIServerLoadBalancerFloatingIP")
+		return nil
+	}
+	if clusterProviderSpec.APIServerLoadBalancerPort == 0 {
+		klog.V(3).Info("No need to create loadbalancer, due to missing APIServerLoadBalancerPort")
+		return nil
+	}
+
+	loadBalancerName := fmt.Sprintf("%s-cluster-%s-%s", networkPrefix, clusterName, kubeapiLBSuffix)
+	klog.Infof("Reconciling loadbalancer %s", loadBalancerName)
+
+	// lb
+	lb, err := checkIfLbExists(s.client, loadBalancerName)
+	if err != nil {
+		return err
+	}
+	if lb == nil {
+		klog.Infof("Creating loadbalancer %s", loadBalancerName)
+		lbCreateOpts := loadbalancers.CreateOpts{
+			Name:        loadBalancerName,
+			VipSubnetID: clusterProviderStatus.Network.Subnet.ID,
+		}
+
+		lb, err = loadbalancers.Create(s.client, lbCreateOpts).Extract()
+		if err != nil {
+			return fmt.Errorf("error creating loadbalancer: %s", err)
+		}
+		err = waitForLoadBalancer(s.client, lb.ID, "ACTIVE")
+		if err != nil {
+			return err
+		}
+	}
+
+	// floating ip
+	fp, err := checkIfFloatingIPExists(s.client, clusterProviderSpec.APIServerLoadBalancerFloatingIP)
+	if err != nil {
+		return err
+	}
+	if fp == nil {
+		klog.Infof("Creating floating ip %s", clusterProviderSpec.APIServerLoadBalancerFloatingIP)
+		fpCreateOpts := &floatingips.CreateOpts{
+			FloatingIP:        clusterProviderSpec.APIServerLoadBalancerFloatingIP,
+			FloatingNetworkID: clusterProviderSpec.ExternalNetworkID,
+			PortID:            lb.VipPortID,
+		}
+		fp, err = floatingips.Create(s.client, fpCreateOpts).Extract()
+		if err != nil {
+			return fmt.Errorf("error allocating floating IP: %s", err)
+		}
+		err = waitForFloatingIP(s.client, fp.ID, "ACTIVE")
+		if err != nil {
+			return err
+		}
+	}
+
+	// lb listener
+	portList := []int{clusterProviderSpec.APIServerLoadBalancerPort}
+	portList = append(portList, clusterProviderSpec.APIServerLoadBalancerAdditionalPorts...)
+	for _, port := range portList {
+		lbPortObjectsName := fmt.Sprintf("%s-%d", loadBalancerName, port)
+
+		listener, err := checkIfListenerExists(s.client, lbPortObjectsName)
+		if err != nil {
+			return err
+		}
+		if listener == nil {
+			klog.Infof("Creating lb listener %s", lbPortObjectsName)
+			listenerCreateOpts := listeners.CreateOpts{
+				Name:           lbPortObjectsName,
+				Protocol:       "TCP",
+				ProtocolPort:   port,
+				LoadbalancerID: lb.ID,
+				//DefaultPoolID:          d.Get("default_pool_id").(string), TODO check vio4/5 // remove before PR
```

Comment:
should this go away? `// remove before PR`

Reply:
Oh yes, good that I noted to remove it *before* PR :)